npps4/webui/endpoints/unlock_backgrounds.py

```diff
@@ -0,0 +1,51 @@
+import urllib.parse
+
+import fastapi
+import sqlalchemy
+
+from ...idol.system import background
+from ... import idol
+from ...db import item
+from ...db import main
+from .. import template
+from ...app import webui
+
+
+async def get_backgrounds(uid: str, to_unlock: list[str]) -> tuple[list[item.Background], list[item.Background]]:
+    async with idol.BasicSchoolIdolContext(lang=idol.Language.en) as context:
+        target_user = await context.db.main.get(main.User, uid)
+
+        if target_user is None:
+            return [], []
+
+        q = sqlalchemy.select(item.Background)
+        result = await context.db.item.execute(q)
+
+        user_backgrounds = await background.get_backgrounds(context, target_user)
+        unlocked_ids = [bg.id for bg in user_backgrounds]
+
+        for bg_id in map(int, to_unlock):
+            if bg_id not in unlocked_ids:
+                await background.unlock_background(context, target_user, bg_id)
+                unlocked_ids.append(bg_id)
+
+        all_backgrounds = result.scalars().all()
+        unlocked_backgrounds: list[item.Background] = [bg for bg in all_backgrounds if bg.background_id in unlocked_ids]
+        locked_backgrounds: list[item.Background] = [
+            bg for bg in all_backgrounds if bg.background_id not in unlocked_ids
+        ]
+
+    return unlocked_backgrounds, locked_backgrounds
```

Comment:
Hmm, I'm bit confused with the function name and what the function does. The funtion name `get_backgrounds` implies this is a "read-only" operation (it will get all unlocked and locked backgrounds of user), but at same time it calls `background.unlock_background` which is "write" operation. Can you explain?

Reply:
You're right this should have been renamed.  Originally I had things split up such that `get_background` was a read-only operation and a separate function did the updating, before I decided it would be simpler to do both in the same context.  I can rename it `do_unlock_backgrounds` unless there's a better name you'd prefer.

Reply:
You can do `await background.unlock_background` inside the `unlock_backgrounds`. Well, everything's clear, I'll merge this ASAP.

Reply:
Great, thanks!

My motivation to separate out the db access into separate methods was again with an eye towards using them for both web and command line interfaces.  Since there wont be command line utilities going forward maybe this should all be folded into `unlock_backgrounds`?